### DISCUSSION TITLE
Question regarding icons

### DISCUSSION THREAD

**Discussion by AlbusPercivalDumbledore** [aminomancer/uc.css.js#35] [General]
Hey, u/motherstylus! How are you doing? This is clueless-nomad from r/firefoxcss. Sorry that I deleted all my posts & account. Was struggling mentally, and really needed a break from reddit. Realized afterwards that shouldn't have deleted the posts regardless :(

I'm using your icons & scripts, and have had an issue since a few updates back: a small black dot appears behind icons included with scripts. Scourged your style-sheets for any relevant rules, but couldn't find one. 

Ex, in Vertical Tabs Pane, behind pin, tabs toggle, and close icons
![1](https://user-images.githubusercontent.com/82368617/157080735-1c6712be-0c16-46d6-b9a7-3f1bac55f989.png)

In all tabs Menu, behind close icon
![2](https://user-images.githubusercontent.com/82368617/157080932-c86b290d-5436-4bc0-b6f8-3e375182c52e.png)

How can these be removed? I've enabled all required prefs enabled in about:config.

**Comment by aminomancer** [comment]
Hi sorry I somehow didn't notice this! I've seen something like that before. Are you still dealing with this issue? Can you show me all the CSS that has the tag `tooltip` in it? I've seen something like this before. I think it is caused by drawing tooltips in the root of SVG documents. Those icons are basically mini documents so I think they technically have native-anonymous tooltips that are just not visible. Or something like that. Might be something in your agent sheet.

**Reply by AlbusPercivalDumbledore** [reply]
Yup, still having the issue, and can confirm tooltip CSS is exactly what's causing it. Changed tooltip background to red, and the dot changed to red as well.

This is what I have in userChrome.as.css:

```
/* Style Tooltips */
tooltip[default][page] {
  -moz-appearance: none !important;
  font-size: 1.05rem !important;
  font-family: "Fira Sans" !important;
  font-weight: normal !important;
  color: #fff !important;
  background-color: #000;
  padding: 5px;
  border-radius: 5px;
  box-shadow: none !important;
  border: 1px solid #181818 !important;
}
```
Pretty sure I stole it from your stylesheets, and modified it a bit, but couldn't find it there now.

**Reply by aminomancer** [reply]
Well here's the [stylesheet](https://github.com/aminomancer/uc.css.js/blob/master/userChrome.ag.css#L417-L444) I use for native anonymous tooltips. And [this sheet](https://github.com/aminomancer/uc.css.js/blob/master/uc-misc.css), which is loaded by regular userChrome.css, has a ton of rules for individual tooltips. (The variables are [in here](https://github.com/aminomancer/uc.css.js/blob/master/resources/layout/uc-low-globals.css), they don't need to be variables, just makes it easier to share between the previous 2 sheets.)

These work a bit differently though, putting the background on the tooltip's child element instead of the tooltip element. I did that in order to support giving the tooltip shadows. The shadow can only be drawn within a popup frame that's defined by the dimensions of the tooltip element, so if we put a shadow on the main `tooltip` element, it won't be visible. But if we give the tooltip a `transparent` background and some padding, and add a background to the next element down (usually the text), then we can add a shadow to that next element too, and it will draw within the now-transparent tooltip popup area.

However, that causes problems for some specific tooltips (the back and forward button tooltips iirc) whose structure is basically:
```
<tooltip>
  <description/>
  <description/>
</tooltip>
```
Since there are 2 child elements (2 rows of text), we can't give them a single shadow or a single background. So I resolve that with [this script](https://github.com/aminomancer/uc.css.js/blob/master/JS/tooltipShadowSupport.uc.js), which wraps the 2 rows of text in a box `.uc-tooltip-box`. That way we can apply the background/shadow/border/etc. rules to the box instead of the tooltip or the descriptions. The script also sets an attribute on the root element so our CSS can check to make sure the script has been applied before applying styles, like `:root[tooltip-shadow-support] tooltip {...}`

Anyway, that's what I currently use for tooltips. But I guess your problem can be mostly resolved without going to such lengths. I think it only happens within SVG icons with a data URI. So, you can wrap the ruleset in a @-moz-document rule like this:
```
@-moz-document regexp("^(?!data).*") {
  tooltip[default="true"][page="true"] {
    ...
  }
}
```
That belongs in your agent sheet (userChrome.as.css or userChrome.ag.css depending on what script you use to load the agent sheet).

In your regular userChrome.css file, you should have the same rule but without the @-moz-document rule or the default and page attributes:
```
tooltip {
  ...
}
```
The attributes and the @-moz-document rule are just intended to make the selector more specific, so that it only targets the one native anonymous tooltip we actually want to target: the one that Firefox uses to draw tooltips for elements that have a `tooltiptext` attribute. So there should only be one per document. Other tooltips are either 1) chrome tooltips, which can already be styled in userChrome.css without needing an agent sheet; or 2) not really supposed to be drawn at all, they're just artifacts of the way Firefox structures document viewers.

The latter case is why you're seeing this bug. Normally those sort of pseudo-tooltips in SVG contexts are invisible because they have no content and aren't open, but they're showing up for you because you're setting `appearance: none` and `background-color` and `padding: 5px`. Each of these on its own wouldn't be a problem, but together they basically code for a 10px black square. So we have to restrict the ruleset to only apply when the tooltip is actually supposed to be visible.

**Reply by AlbusPercivalDumbledore** [reply]
The regex filter didn't work unfortunately, so I tried using tooltipShadowSupport script with my CSS. That still caused problems for some tooltips, like you had mentioned (Reload button, for one. Tooltip was appearing inside another container). No amount of trial & error helped to get rid of it.

So, I wound up using the script and your rules. Only had to modify the background colors. That solved the issue. 

Thanks a lot MS!

**Reply by aminomancer** [reply]
It's regexp in CSS, with a p. Well, it worked for me. I tested it with your stylesheet

**Reply by aminomancer** [reply]
And oh, the script alone doesn't have anything to do with CSS. It doesn't do anything except affect a couple tooltips. Like I said, it just wraps them in a little wrapper. It's the CSS rules that matter since they put the background on the child element, instead of the tooltip element. The fake tooltips in the icons don't have any child element so they end up fully transparent

**Reply by AlbusPercivalDumbledore** [reply]
I copied over the @-moz-document rule you gave, so 'regexp' did have 'p'. Also made sure to have [default="true"][page="true"]. 

I'm at my wits end trying to get it work that way. Since you said it worked for you, I've tried every which way, but it just doesn't work. Checked the rules with Prettier for any mistakes; none.

I'm clutching at straws here, but could it have anything to do with me being on Linux? Is there any chance that regexp works differently on Linux? Or anything to do with encoding of the file? 

I found this thread, if its of any interest: https://serverfault.com/questions/77686/why-does-this-regex-not-work-on-linux

**Reply by aminomancer** [reply]
No I don't think it has anything to do with operating system, since this is just a very simple negative lookahead. I would be really shocked if it didn't work on linux. How do you know it's not working? What icons are you still seeing a black box in? Does the @-moz-document rule fix some of them but not all, or is there no difference at all? Also, what happens if you delete the entire rule?

**Reply by AlbusPercivalDumbledore** [reply]
All the ones where it was originally visible. VerticalTabsPane icon, Pin icon, Close icon in All Tabs Menu, Icons in Hamburger menu that I'd copied over from your style-sheets. 

Using/ not using @moz-document-rule doesn't make any difference. And, if entire rule (including tooltip{..}?) is removed, tooltips use system theme (GTK theme on Linux).

**Reply by aminomancer** [reply]
What script are you using to load the agent sheet? Maybe you should switch to using [my script](https://github.com/aminomancer/uc.css.js/blob/master/JS/userChromeAgentAuthorSheetLoader.uc.js). Are you using the latest version of [fx-autoconfig](https://github.com/MrOtherGuy/fx-autoconfig)? Otherwise, have you tried changing it to `regexp("^\(?!data\).*")`?

**Reply by AlbusPercivalDumbledore** [reply]
Was using userChrome_as_css_module.uc.js that comes with fx-autoconfig. No change even after switching to your script, updating fx-autoconfig, and changing regexp. 

Lets bury this here, haha. Using your stylesheet gives the same result, without the whole shebang.

**Reply by AlbusPercivalDumbledore** [reply]
Hey, u/MotherStylus. I have a question not related to firefox css. Some websites have text directly included in html. Ex, 

```
<span> Unwanted Text. That I do not (NOT) want to display. </span>
<span> Important Text  </span>
```
Would it be possible to remove such unwanted text, without touching Important Text?

I searched around, and found that it could be done with a content script:

```
function extractContent(s)
{
var span= document.createElement('span');
  span.innerHTML= s;
var res=span.innerText;
     if(res.match("Unwanted Text. That I do not (NOT) want to display."))
       {
        res= res.replace("Unwanted Text. That I do not (NOT) want to display.","");
       }
  return res;
}

```
Yet, this doesn't hide the text.

**Reply by aminomancer** [reply]
I'm not sure I follow what you're trying to do. What about the text makes it unwanted? It looks the same except for the string inside it. But then what are the unwanted strings?

The function you pasted doesn't really do anything. It creates an empty span (not placed in the tree, that has to be done with DOMNode.appendChild or similar) and sets its innerHTML to `s`, which is going to make its innerText `s`. Doesn't seem to interact at all with the document. I thought you were hoping to find text strings in the document and hide them if they match some chosen parameters?

Unfortunately this kind of thing is very slow. It has to iterate over all the content in the document. You can actually do this with uBlock Origin by adding a rule like `example.com,domain.com##span:has-text(Unwanted Text. That I do not \(NOT\) want to display.)`

As for a content script, you can do this [XPath](https://developer.mozilla.org/en-US/docs/Web/XPath) with a violentmonkey script (again, it's pretty slow)...
```
// ==UserScript==
// @name           XPath Content Hider
// @version        1.0
// @author         aminomancer
// @homepageURL    https://github.com/aminomancer/uc.css.js
// @description    Hide nodes with chosen properties defined in const options below.
// @supportURL     https://developer.mozilla.org/en-US/docs/Web/XPath
// @run-at         document-body
// @grant          none
// @match          http*://example.com/*
// ==/UserScript==

/**
 * @typedef XPathExpressionOptions
 * @property {string} textContent match text content.
 * @property {string} tagName match node tag name.
 * @property {string} parentElementTagName match node parent's tag name.
 */
const options = {
  textContent: "Unwanted Text. That I do not (NOT) want to display.",
  tagName: "span",
  parentElementTagName: null,
};

/**
 * Remove text and repeat on DOMContentLoaded and load events.
 * @param {XPathExpressionOptions} exp search options.
 */
function setupHandlers(exp) {
  let hide = () => hideText(exp.textContent, exp.tagName, exp.parentElementTagName);
  document.addEventListener("DOMContentLoaded", hide, { once: true });
  window.addEventListener("load", hide, { once: true });
  hide();
}

/**
 * Find matching nodes with XPathEvaluator.evaluate and hide them.
 * @param {XPathExpressionOptions} exp search options.
 */
async function hideText(exp) {
  let xpe = new XPathEvaluator();
  let nsResolver = xpe.createNSResolver(document.documentElement);
  let { textContent, tagName = "*", parentElementTagName } = exp;
  let p = parentElementTagName ? `${parentElementTagName}/child::` : "";
  let result = xpe.evaluate(
    `//${p}${tagName}[text()="${textContent}"]`,
    document,
    nsResolver,
    XPathResult.UNORDERED_NODE_ITERATOR_TYPE,
    null
  );
  let found = [];
  let res;
  while ((res = result.iterateNext())) found.push(res);
  if (found.length > 0) {
    found.forEach(node => node?.style.setProperty("display", "none", "important"));
  }
}

setupHandlers(options);
```

**Reply by AlbusPercivalDumbledore** [reply]
Its a website providing content for an exam I'm studying for. It was full of banner ads for its own offerings, which I could hide with uBO. But, recently they started including 4-5 lines long text ads, running across the width of website, in 4 times the normal font size, in every post. This text didn't have any unique selectors, so hiding it got rid of the relevant content too. That's why 'Unwanted' and 'Important' text :D

I found a solution somewhere suggesting `:has-text`, but was trying to use it in userContent.css, where it wasn't having any effect. The script I pasted had an alert function as well, which was giving a popup if test matching that string was found. So, it seemed to be working, yet wasn't hiding the text. And, I'm stupid with scripts, so thought of asking you if anything was amiss.

Didn't even think of using uBO. Now, its hidden. Also, could hide other irrelevant content with uBO, so using uBO, instead of content script (Thankyou for taking the time to write it).

Thanks a ton! :)
